File: app/src/main/java/com/ufkoku/demo_app/ui/view/DataView.java
package com.ufkoku.demo_app.ui.view;

import android.content.Context;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.util.AttributeSet;
import android.view.View;
import android.widget.FrameLayout;

import com.ufkoku.demo_app.R;
import com.ufkoku.demo_app.entity.AwesomeEntity;
import com.ufkoku.demo_app.ui.view.adapter.DataAdapter;

import java.util.List;

public class DataView extends FrameLayout implements IDataView {

    private ViewListener listener;

    private RecyclerView recyclerView;

    private View vWaitView;

    public DataView(Context context) {
        super(context);
    }

    public DataView(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    public DataView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }

    @Override
    protected void onFinishInflate() {
        super.onFinishInflate();

        recyclerView = (RecyclerView) findViewById(R.id.recycler);
        recyclerView.setLayoutManager(new LinearLayoutManager(getContext(), LinearLayoutManager.VERTICAL, false));

        vWaitView = findViewById(R.id.waitView);

        findViewById(R.id.retainable).setOnClickListener(view -> {
            if (listener != null) {
                listener.onRetainableClicked();
            }
        });
        findViewById(R.id.savable).setOnClickListener(view -> {
            if (listener != null) {
                listener.onSavableClicked();
            }
        });
    }

    public void setListener(ViewListener listener) {
        this.listener = listener;
    }

    @Override
    public void populateData(final List<AwesomeEntity> entities) {
        recyclerView.post(new Runnable() {
            @Override
            public void run() {
                recyclerView.setAdapter(new DataAdapter(getContext(), entities));
            }
        });
    }

    @Override
    public void setWaitViewVisible(boolean visible) {
        vWaitView.setVisibility(visible ? VISIBLE : GONE);
    }

    public interface ViewListener {

        void onRetainableClicked();

        void onSavableClicked();

    }

}

File: app/src/main/kotlin/com/ufkoku/demo_app/ui/view/DataView.kt
package com.ufkoku.demo_app.ui.view

import android.content.Context
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.util.AttributeSet
import android.view.View
import android.widget.FrameLayout

import com.ufkoku.demo_app.R
import com.ufkoku.demo_app.entity.AwesomeEntity
import com.ufkoku.demo_app.ui.view.adapter.DataAdapter

class DataView : FrameLayout, DataAdapter.AdapterListener {

    private var listener: ViewListener? = null

    private var recyclerView: RecyclerView? = null

    private var vWaitView: View? = null

    constructor(context: Context) : super(context) {}

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {}

    constructor(context: Context, attrs: AttributeSet, defStyleAttr: Int) : super(context, attrs, defStyleAttr) {}

    override fun onFinishInflate() {
        super.onFinishInflate()

        recyclerView = findViewById<View>(R.id.recycler) as RecyclerView
        recyclerView!!.layoutManager = LinearLayoutManager(context, LinearLayoutManager.VERTICAL, false)

        vWaitView = findViewById(R.id.waitView)

        findViewById<View>(R.id.retainable).setOnClickListener { view ->
            if (listener != null) {
                listener!!.onRetainableClicked()
            }
        }
        findViewById<View>(R.id.savable).setOnClickListener { view ->
            if (listener != null) {
                listener!!.onSavableClicked()
            }
        }
    }

    fun setListener(listener: ViewListener) {
        this.listener = listener
    }

    fun populateData(entities: List<AwesomeEntity>) {
        recyclerView!!.post {
            val adapter = DataAdapter(context, entities)
            adapter.listener = this
            recyclerView!!.adapter = adapter
        }
    }

    fun setWaitViewVisible(visible: Boolean) {
        vWaitView!!.visibility = if (visible) View.VISIBLE else View.GONE
    }

    override fun onItemClicked(entity: AwesomeEntity) {
        listener?.onItemClicked(entity)
    }

    interface ViewListener {

        fun onItemClicked(entity: AwesomeEntity)

        fun onRetainableClicked()

        fun onSavableClicked()

    }

}
